packages/react/src/Button/ButtonBase.tsx

```diff
@@ -84,6 +89,226 @@ const ButtonBase = forwardRef(
+                    {
+                      /* Render a leading visual unless the button is in a loading state.
+                     Then replace the leading visual with a loading spinner. */
+                      LeadingVisual && renderVisual(LeadingVisual, Boolean(loading), 'leadingVisual')
+                    }
+                    {children && (
+                      <span data-component="text" id={loading ? `${uuid}-label` : undefined}>
+                        {children}
+                      </span>
+                    )}
+                    {
+                      /* If there is a count, render a counter label unless there is a trailing visual.
+                     Then render the counter label as a trailing visual.
+                     Replace the counter label or the trailing visual with a loading spinner if:
+                     - the button is in a loading state
+                     - there is no leading visual to replace with a loading spinner
+                  */
+                      count !== undefined && !TrailingVisual
+                        ? renderVisual(
+                            () => <CounterLabel data-component="ButtonCounter">{count}</CounterLabel>,
+                            Boolean(loading) && !LeadingVisual,
+                            'trailingVisual',
+                          )
+                        : TrailingVisual
+                        ? renderVisual(TrailingVisual, Boolean(loading) && !LeadingVisual, 'trailingVisual')
+                        : null
+                    }
+                  </Box>
+                  {
+                    /* If there is a trailing action, render it unless the button is in a loading state
+                   and there is no leading or trailing visual to replace with a loading spinner. */
+                    TrailingAction &&
+                      renderVisual(
+                        TrailingAction,
+                        Boolean(loading) && !LeadingVisual && !TrailingVisual,
+                        'trailingAction',
+                      )
+                  }
+                </>
+              )}
+            </StyledButton>
+            {loading && (
+              <VisuallyHidden>
+                <AriaStatus id={loadingAnnouncementID}>{loadingAnnouncement}</AriaStatus>
+              </VisuallyHidden>
+            )}
+          </ConditionalWrapper>
+        )
+      }
+      return (
+        <ConditionalWrapper
+          // If anything is passsed to `loading`, we need the wrapper:
+          // If we just checked for `loading` as a boolean, the wrapper wouldn't be rendered
+          // when `loading` is `false`.
+          // Then, the component re-renders in a way that the button will lose focus when switching between loading states.
+          if={typeof loading !== 'undefined'}
+          sx={{display: block ? 'block' : 'inline-block'}}
+          data-loading-wrapper
+        >
+          <StyledButton
+            as={Component}
```

Comment:
```suggestion
          <Component
```

I don't think we need to render the StyledButton here